src/main/kotlin/store.kt

```diff
@@ -0,0 +1,101 @@
+import java.util.*
+
+open class StoreException: Exception()
+class KeyNotFound(val key: String): StoreException()
+class NoTransaction: StoreException()
+
+sealed class Element
+data class Value(val value: String): Element()
+object Deleted: Element()
+
+class Store {
+    // the store tracks a stack of Elements per key
+    // the current value is always the first element on the stack
+    // so reads never need to walk the stack
+    private val store: MutableMap<String, Deque<Element>> = mutableMapOf()
+    // the stack tracks a set of keys per tx level
+    // so that we don't have to search all keys on abort/commit
```

Comment:
This whole solution is brilliant. I keep thinking about it, going over how it works. I will probably try to reimplement it using slightly different code, just to prove to myself that I really get it.